scripts/west_commands/zephyr_ext_common.py

```diff
@@ -49,17 +48,63 @@ def check_force(self, cond, msg):
             log.die('refusing to proceed without --force due to above error')
 
 
-def cached_runner_config(build_dir, cache):
-    '''Parse the RunnerConfig from a build directory and CMake Cache.'''
-    board_dir = cache['ZEPHYR_RUNNER_CONFIG_BOARD_DIR']
-    elf_file = cache.get('ZEPHYR_RUNNER_CONFIG_KERNEL_ELF')
-    hex_file = cache.get('ZEPHYR_RUNNER_CONFIG_KERNEL_HEX')
-    bin_file = cache.get('ZEPHYR_RUNNER_CONFIG_KERNEL_BIN')
-    gdb = cache.get('ZEPHYR_RUNNER_CONFIG_GDB')
-    openocd = cache.get('ZEPHYR_RUNNER_CONFIG_OPENOCD')
-    openocd_search = cache.get('ZEPHYR_RUNNER_CONFIG_OPENOCD_SEARCH')
-
-    return RunnerConfig(build_dir, board_dir,
-                        elf_file, hex_file, bin_file,
-                        gdb=gdb, openocd=openocd,
-                        openocd_search=openocd_search)
+def load_dot_config(path):
```

Comment:
From the look of it, I like the switch from CMake cache to `.config`.
But I'm concerned that there is a hole in the loop.

If a user run's `menuconfig` without rebuilding, then the `.config` could contain new values that are not matching the current build.

In the old system, the CMakeCache would still match the values from last build, and even if user does `menuconfig` those CMake cache values stays the same (and thus they matches the last `ninja` build result).
Not until the user does a new `ninja` build, then CMake is re-run and the cache values are updated to match the updated `.config` value, and the same is true for build output/result.
This means CMake cache values stays consistent with the last build output.

Reply:
I agree this is a change in behavior, but I'm not sure what you mean by "hole in the loop".

What change are you asking for?

Reply:
I guess what I'm saying is that I consider the difference you are describing at the level of implementation details.

In the current implementation, the user could run `cmake-gui` and update the internals and get different `west sign` results. In the PR, the user can run `ninja menuconfig` and do the same. What is the desired behavior you are hoping to see? I think *some* behavior changes following incremental builds which tweak things are necessary either way.

Reply:
> I guess what I'm saying is that I consider the difference you are describing at the level of implementation details.
> 
> In the current implementation, the user could run `cmake-gui` and update the internals and get different `west sign` results. 

Except that those setting have `  set(... CACHE STRING "..." FORCE)` so changes will be overwritten.
But now that you mention `cmake-gui, those settings should actually have been `INTERNAL`, but guess it doesn't make much sense to change that now.

> In the PR, the user can run `ninja menuconfig` and do the same.

`menuconfig` is the standard way to configure Zephyr, and `cmake-gui` is not, so the chance that someone change those settings in `menuconfig` and thus get this out-of-sync is much higher (compared to `cmake-gui`).

> What is the desired behavior you are hoping to see? I think _some_ behavior changes following incremental builds which tweak things are necessary either way.

I'm not sure, that's why I was raising it, to see if there where any good ideas.
One idea could be to check if the settings are out-of-sync, as minimum, to check if the file(s) referenced exists.
And if they don't, report to the user that a re-build is required.
But i'm unsure if there could be situations where changed settings still requires a rebuild, even if files are present.

I looked into `cmake --build <build-folder> -- -n` for dry run, but that only work as expected for `ninja` and not for `make`.
For make:
```
No-op”. Causes make to print the recipes that are needed to make the targets up to date, but not actually execute them.
Note that some recipes are still executed, even with this flag 
```

So other ideas are welcome as well.

Reply:
> So other ideas are welcome as well.

I think we may actually regard this as a bug fix: `west help sign` says:

```
The image header size, alignment, and slot sizes are determined from
the build directory using .config and the device tree.
```

So:

1. this PR is making the implementation match the command help
2. the help doesn't mention updating the binary if you change .config

I therefore believe this should remain as-is.

Reply:
> ```
> The image header size, alignment, and slot sizes are determined from
> the build directory using .config and the device tree.
> ```

Why is image header size, alignment, and slot sizes relevant in this context ?
I'm talking about the CMake variables `ZEPHYR_RUNNER_CONFIG_KERNEL_ELF`, `ZEPHYR_RUNNER_CONFIG_KERNEL_BIN`, and similar that you are changing with this PR.
Those names have nothing to do with the settings you refer to above.

Let's my try to show you why I think there is now a loophole when taking this PR as-is:

On `zephyr/master`, if I can do the following:
```
$ rm -rf build; cmake -DBOARD=nrf52840dk_nrf52840 -GNinja -DCONFIG_BOOTLOADER_MCUBOOT=y -Bbuild  .
# cmake running

$ ninja -Cbuild 
# ninja building
[131/131] Linking C executable zephyr/zephyr.elf

$ west -v sign -t imgtool -- --key privatekey.pem  
=== signed binaries:
bin: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.bin /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
hex: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.hex /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex

# Let's update the output name from "zephyr" --> "myproduct", similar to using menuconfig.
$ sed 's/\(CONFIG_KERNEL_BIN_NAME="\).*\("\)/\1myproduct\2/g' build/zephyr/.config > tmpfile; mv tmpfile build/zephyr/.config

$ west -v sign -t imgtool -- --key privatekey.pem  
# Note: still using output names from LAST build
=== signed binaries:
bin: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.bin /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
hex: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.hex /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex

$ ninja -Cbuild 
# ninja rebuilding.
[126/126] Linking C executable zephyr/myproduct.elf

$ west -v sign -t imgtool -- --key privatekey.pem
# Now using names from NEW build
bin: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/myproduct.bin /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
hex: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/myproduct.hex /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
$
tora@ubuntu /p/g/n/zephyr> cat 27441_comments_sign_existing.txt 
$ rm -rf build; cmake -DBOARD=nrf52840dk_nrf52840 -GNinja -DCONFIG_BOOTLOADER_MCUBOOT=y -Bbuild  .
# cmake running

$ ninja -Cbuild 
# ninja running
[131/131] Linking C executable zephyr/zephyr.elf

$ west -v sign -t imgtool -- --key privatekey.pem  
=== signed binaries:
bin: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.bin /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
hex: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.hex /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex

# Let's update the output name from "zephyr" --> "myproduct", similar to using menuconfig.
$ sed 's/\(CONFIG_KERNEL_BIN_NAME="\).*\("\)/\1myproduct\2/g' build/zephyr/.config > tmpfile; mv tmpfile build/zephyr/.config

$ west -v sign -t imgtool -- --key privatekey.pem  
# Note: still using output names from LAST build
=== signed binaries:
bin: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.bin /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
hex: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.hex /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex

$ ninja -Cbuild 
# ninja rebuilding.
[126/126] Linking C executable zephyr/myproduct.elf

$ west -v sign -t imgtool -- --key privatekey.pem
# Now using names from NEW build
bin: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/myproduct.bin /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
hex: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
/home/tora/.local/bin/imgtool sign --version 0.0.0+0 --align 4 --header-size 512 --slot-size 421888 --key privatekey.pem /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/myproduct.hex /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
$
```
and as seen, `west sign` will keep signing the LAST build result, until a rebuild is done, even if I modify `.config`.
The rebuild determines when the signed file is changed.

This is aligned with how `west flash` works:
```
$ rm -rf build; cmake -DBOARD=nrf52840dk_nrf52840 -GNinja -DCONFIG_BOOTLOADER_MCUBOOT=y -Bbuild  .
$ ninja -Cbuild 
[131/131] Linking C executable zephyr/zephyr.elf
$ west flash --skip-rebuild 
-- west flash: using runner nrfjprog
-- runners.nrfjprog: Flashing file: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.hex

# Let's update the output name from "zephyr" --> "myproduct", similar to using menuconfig.
$ sed 's/\(CONFIG_KERNEL_BIN_NAME="\).*\("\)/\1myproduct\2/g' build/zephyr/.config > tmpfile; mv tmpfile build/zephyr/.config

$ west flash --skip-rebuild 
# Note: still flashing output from LAST build
-- west flash: using runner nrfjprog
-- runners.nrfjprog: Flashing file: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.hex

$ ninja -Cbuild 
# ninja rebuilding.
[126/126] Linking C executable zephyr/myproduct.elf

$ west flash --skip-rebuild
# Now flashing output from NEW build
-- west flash: using runner nrfjprog
-- runners.nrfjprog: Flashing file: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/myproduct.hex
```
`west flash` will keep flashing the LAST build result, and if the filename is changed, then that filename will we be used AFTER a rebuild. Until the rebuild, the previous file name is still used.

So, `west flash` and `west sign` behaves the same (and so does `west debug`) on `zephyr/master`.

With this PR, `west sign` changes behavior, so if I try doing the same with this PR, then I get:
```
$ rm -rf build; cmake -DBOARD=nrf52840dk_nrf52840 -GNinja -DCONFIG_BOOTLOADER_MCUBOOT=y -Bbuild  .
# cmake running

$ ninja -Cbuild 
# ninja running
[131/131] Linking C executable zephyr/zephyr.elf

$ west sign -t imgtool -- --key privatekey.pem  
=== signing binaries
unsigned bin: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.bin
signed bin:   /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
unsigned hex: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.hex
signed hex:   /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex

# Let's update the output name from "zephyr" --> "myproduct", similar to using menuconfig.
$ sed 's/\(CONFIG_KERNEL_BIN_NAME="\).*\("\)/\1myproduct\2/g' build/zephyr/.config > tmpfile; mv tmpfile build/zephyr/.config

# ====== FAILURE BELOW =======
$ west sign -t imgtool -- --key privatekey.pem  
# Trying to use names from future build, and fails.
FATAL ERROR: no unsigned .bin found at /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/myproduct.bin

$ ninja -Cbuild 
# ninja rebuilding.
[126/126] Linking C executable zephyr/myproduct.elf

$ west sign -t imgtool -- --key privatekey.pem  
# After a rebuild, signing works
=== signing binaries
unsigned bin: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/myproduct.bin
signed bin:   /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.bin
unsigned hex: /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/myproduct.hex
signed hex:   /projects/github/ncs/zephyr/samples/hello_world/build/zephyr/zephyr.signed.hex
```
as you notice, if I change `CONFIG_KERNEL_BIN_NAME` without rebuilding, then I get a failure in `west sign` cause it will try signing the NEW image which haven't been built, instead of signing the LAST build image.

This is also different from what `west flash` does.
`west flash` still behaves the same as in `zephyr/master`.

Reply:
How about using the information from `runners.yaml` for signing ?